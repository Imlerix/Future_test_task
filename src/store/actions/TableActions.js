import * as types from '../../constants/actionTypes';
import axios from 'axios'
const url = `${process.env.REACT_APP_SERVERURL}`;

export function getRows(dataSize = 32) {
    return (dispatch, getState) => {

        const params = {
            rows: dataSize,
            id: '{number|1000}',
            firstName: '{firstName}',
            lastName: '{lastName}',
            email: '{email}',
            phone: '{phone|(xxx)xxx-xx-xx}',
            address: '{addressObject}',
            description: '{lorem|32}',
        }
        if (dataSize > 32) params.delay = 3;

        axios.get(url, { params })
            .then(res => {
                let data = {
                    type: types.GET_ROWS_FROM_SERVER,
                    rows: res.data,
                }

                dispatch(data);
            })
            .catch(err => {
                dispatch([])
                console.error(err)
            })
    }
}

export function selectRow(userObj) {
    return {
        type: types.SELECT_USER_IN_TABLE,
        user: userObj
    }
}

export function addUserInTable(userObj) {
    return {
        type: types.ADD_USER_IN_TABLE,
        user: userObj
    }
}
