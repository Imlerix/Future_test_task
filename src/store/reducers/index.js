export { default as table } from './TableReducer';
