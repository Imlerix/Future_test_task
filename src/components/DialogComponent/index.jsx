import React from 'react';
import style from './style.module.css';
import PropTypes from 'prop-types';
import {connect} from 'react-redux'
import { bindActionCreators } from 'redux';
import * as TableActions from '../../store/actions/TableActions';

import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import ListItem from "@material-ui/core/ListItem";
import List from "@material-ui/core/List";
import Button from "@material-ui/core/Button";
import SaveIcon from '@material-ui/icons/Save';
import TextField from "@material-ui/core/TextField";
import Tooltip from "@material-ui/core/Tooltip";


class DialogComponent extends React.Component{
    constructor(props){
        super(props)
        this.state = {
            isAllInputsFull: false,
            fields: [
                {
                    name: 'id',
                    label: 'ID',
                    regExp: /^\d+$/gi,
                    isCorrect: true,
                    hint: 'Только цифры',
                    value: ''
                },
                {
                    name: 'firstName',
                    label: 'Имя',
                    regExp: /^[a-zA-Zа-яА-Я]+$/ig,
                    isCorrect: true,
                    hint: 'Только буквы без пробелов',
                    value: ''
                },
                {
                    name: 'lastName',
                    label: 'Фамилия',
                    regExp: /^[a-zA-Zа-яА-Я]+$/ig,
                    isCorrect: true,
                    hint: 'Только буквы без пробелов',
                    value: ''
                },
                {
                    name: 'email',
                    label: 'Эл. почта',
                    regExp: /^(([^<>()\[\]\.,;:\s@\"]+(\.[^<>()\[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/gi,
                    isCorrect: true,
                    hint: 'Например: qwerty@gmail.com',
                    value: ''
                },
                {
                    name: 'phone',
                    label: 'Моб. телефон',
                    regExp: /^((\+7|7|8)+([0-9]){10})$/gi,
                    isCorrect: true,
                    hint: 'Например +79991235678 или 89991235678',
                    value: ''
                },
            ]
        }
    }
    TableActions = bindActionCreators(TableActions, this.props.dispatch);

    checkFormFull = (fields) => {
        const countCorrectFields = fields.reduce((sum, curr) => {
            if (curr.isCorrect && curr.value.length > 0) return ++sum;
        }, 0)
        if (countCorrectFields === fields.length){
            this.setState({isAllInputsFull: true});
        }else{
            this.setState({isAllInputsFull: false});
        }
    }

    onChangeForm = (e) => {
        const index = e.target.id;
        const value = e.target.value;
        const name = e.target.name;

        this.setState((state) => {
            if (value.match(state.fields[index].regExp) || value === '' || value === undefined) {
                state.fields[index].isCorrect = true;
                if (name === 'id'){
                    for (let user of this.props.table.usersRows){
                        if (user.id == value) {
                            state.fields[index].isCorrect = false;
                            state.fields[index].hint = 'Такой ID уже существует';
                            break;
                        }else{
                            state.fields[index].isCorrect = true;
                            state.fields[index].hint = 'Только цифры';
                        }
                    }
                }
            }else{
                state.fields[index].isCorrect = false;
            }
            state.fields[index].value = value;

            this.checkFormFull(state.fields);
            return state;
        });
    }

    handleSave = () => {
        const userObj = this.state.fields.reduce((summaryObj, field) => {
            summaryObj[field.name] = field.value;
            return summaryObj;
        }, {})
        this.setState((state) => {
            state.fields = state.fields.map(field => {
                field.value = '';
                return field;
            })
            return state;
        })

        this.TableActions.addUserInTable(userObj);
        this.handleClose();
    }

    handleClose = () => {
        this.props.onClose();
    };

    render() {
        const {
            open
        } = this.props;

        return(
            <Dialog onClose={this.handleClose}
                    aria-labelledby="simple-dialog-title"
                    open={open}>
                    <DialogTitle>Добавить нового пользователя</DialogTitle>
                    <List>
                        {this.state.fields.map((field, index) => (
                            <ListItem key={index}>
                                <TextField fullWidth
                                           error={!field.isCorrect}
                                           onChange={this.onChangeForm}
                                           helperText={field.hint}
                                           id={index}
                                           name={field.name}
                                           value={field.value}
                                           label={field.label} />
                            </ListItem>
                        ))}
                    </List>
                    <Tooltip disableHoverListener={this.state.isAllInputsFull}
                             title="Сначала заполните поля!"
                             style={{width: '100%'}}
                             aria-label="save">
                        <div style={{width: '100%'}}>
                        <Button variant="contained"
                                fullWidth
                                onClick={this.handleSave}
                                disabled={!this.state.isAllInputsFull}
                                color="primary"
                                size="large"
                                startIcon={<SaveIcon />}>
                                Добавить в таблицу
                        </Button>
                        </div>
                    </Tooltip>
            </Dialog>
        )
    }
}

DialogComponent.propTypes = {
    onClose: PropTypes.func.isRequired,
    open: PropTypes.bool.isRequired,
};

const mapStateToProps = (state) => ({
    table: state.table
})
export default connect(mapStateToProps)(DialogComponent);
