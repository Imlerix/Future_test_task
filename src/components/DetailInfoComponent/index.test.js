import renderer from 'react-test-renderer'
import React from 'react';
import { Provider } from 'react-redux';
import { initStore } from '../../store';
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import EnzymeToJson from 'enzyme-to-json';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import DetailInfoComponentClever, {DetailInfoComponent} from './index';
import * as TableActions from '../../store/actions/TableActions';
import {DialogComponent} from "../DialogComponent";

Enzyme.configure({ adapter: new Adapter() });

const selectedRow = {
    id: 1,
    firstName: 'Ivan',
    lastName: 'Ivanov',
    description: 'Desc',
    phone: '+79991234567',
    address: {
        streetAddress: 'arbatskaya 100',
        city: 'Moskow',
        state: 'Moskow obl',
        zip: '123',
    },
}

const initialState = {
    table: {
        selectedRow
    }
}

let container = null;
const mockStore = initStore();
mockStore.dispatch(TableActions.selectRow(selectedRow));

beforeEach(() => {
    // подготавливаем DOM-элемент, куда будем рендерить
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    // подчищаем после завершения
    unmountComponentAtNode(container);
    container.remove();
    container = null;
});

it("renders correctly", () => {
    act(() => {
        render(
            <Provider store={mockStore}>
                <DetailInfoComponentClever />
            </Provider>
            ,container
        );
    });

    const textContent =
        `Выбран пользователь: ${selectedRow.firstName} ${selectedRow.lastName}`+
        `Описание: ${selectedRow.description}`+
        `Адрес проживания: ${selectedRow.address.streetAddress}`+
        `Город: ${selectedRow.address.city}`+
        `Провинция/штат: ${selectedRow.address.state}`+
        `Индекс: ${selectedRow.address.zip}`+
        `Скрыть детальную информацию`;

    // expect(EnzymeToJson(subject)).toMatchSnapshot();
    expect(container.textContent).toBe(textContent);
});

it("SNAPSHOT renders correctly", () => {
    const subject = shallow(
            <DetailInfoComponent table={{selectedUserRow: selectedRow}}/>
            ,container
    );
    expect(EnzymeToJson(subject)).toMatchSnapshot();
});
