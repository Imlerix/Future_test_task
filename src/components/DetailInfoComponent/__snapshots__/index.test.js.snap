// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOT renders correctly 1`] = `
<div
  className="DetailInfoComponentContainer"
>
  <WithStyles(ForwardRef(Card))>
    <WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(Typography))
        component="h2"
        gutterBottom={true}
        variant="h5"
      >
        Выбран пользователь: 
        Ivan
         
        Ivanov
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        color="textSecondary"
        component="p"
        variant="body2"
      >
        <b>
          Описание: 
        </b>
        Desc
        <br />
        <br />
        <b>
          Адрес проживания: 
        </b>
        arbatskaya 100
        <br />
        <b>
          Город: 
        </b>
        Moskow
        <br />
        <b>
          Провинция/штат: 
        </b>
        Moskow obl
        <br />
        <b>
          Индекс: 
        </b>
        123
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(CardContent))>
    <WithStyles(ForwardRef(CardActions))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[Function]}
        size="small"
      >
        Скрыть детальную информацию
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(CardActions))>
  </WithStyles(ForwardRef(Card))>
</div>
`;
